merge_dynamic_streams_block.py

```diff
@@ -0,0 +1,109 @@
+
+from nio.block.mixins.group_by.group_by import GroupBy
+
+
+@discoverable
+class MergeDynamicStreams(GroupBy, Block):
+
+    """ Merge a dynamic number of streams and combine signals together. """
+
+    ttl = TimeDeltaProperty(default={})
+    notify_once = BoolProperty(default=True)
+    stream = Property(name="Stream", default=None, allow_none=True)
+    version = VersionProperty('0.1.0')
+
+    def __init__(self):
+        super().__init__()
+        # Keep track of the last signal for each stream in each group.
+        # Signals are removed from here when ttl expires.
+        # Stream names are never removed.
+        self._last_signal = defaultdict(lambda: defaultdict(lambda: None))
+        # Keep track of the last signal for each stream in each group.
+        # Signals are removed from here when ttl expires and when notified and
+        # notify_once is True. Stream names are never removed.
+        self._signals = defaultdict(lambda: defaultdict(lambda: None))
+        # Store jobs that remove signals after ttl.
+        self._expiration_jobs = defaultdict(lambda: defaultdict(lambda: None))
+
+    def process_group_signals(self, signals, group, input_id):
+        """Return the merged signals to be notified for this group"""
+        output_signals = []
+        # Group signals into streams and then process each one at a time
+        streams = defaultdict(list)
+        for signal in signals:
+            streams[self.stream(signal)].append(signal)
+        for stream in streams:
+            output_signals.extend(
+                self.process_stream_signals(streams[stream], group, stream))
+        return output_signals
+
+    def process_stream_signals(self, signals, group, stream):
+        """Return the merged signals to be notified for this stream"""
+        merged_signals = []
+        for signal in signals:
+            new_stream = False
+            if stream not in self._last_signal[group]:
+                # Trigger merge streams if this is a new one
+                new_stream = True
+            # Save new signal
+            self._signals[group][stream] = signal
+            self._last_signal[group][stream] = signal
+            # Merge all streams for this group
+            merged_signals.extend(self._merge_streams(group, new_stream))
+        if self.ttl():
+            self._schedule_signal_expiration_job(group, stream)
+        return merged_signals
+
+    def _merge_streams(self, group, new_stream):
+        """Merge signals from each stream and clear if notify once."""
+        if new_stream and self.notify_once():
+            # Use last signal signal even if it was already notified
+            streams = self._last_signal[group]
+        else:
+            # Otherwise use only the signals still available
+            streams = self._signals[group]
+        if not [streams[s] for s in streams if streams[s] is None]:
+            # Only merge streams if we have a signal for every stream
```

Comment:
Per @hansmosh, this conditional is correct, the notify_once check won't remove the signals from the stream